lib/bundle/commands/cleanup.rb

```diff
@@ -145,7 +145,9 @@ def recursive_dependencies(current_formulae, formulae_names, top_level: true)
 
       def taps_to_untap(global: false, file: nil)
         @dsl ||= Brewfile.read(global:, file:)
+        kept_formulae = @dsl.entries.select { |e| e.type == :brew }.map(&:name).map(&Formulary.method(:factory))
```

Comment:
A `kept_formula` method that extracts/shares some logic with `formulae_to_uninstall` and memoizes to avoid doing the same logic twice would be ideal.

Reply:
Happy to make the change but looking at `kept_casks`, it accepts two arguments and then ignores them after the first invocation. Maybe not an issue but seems like a potential footgun.

How much do you want the new method to include? All the resolution of dependencies too? I'm not sure if that would potentially yield more taps and therefore be necessary to make my chance catch all cases.

`@dsl.entries.select { |e| e.type == :brew }.map(&:name)` could arguably be a helper method on `Bundle::Dsl` also.

Reply:
> How much do you want the new method to include? All the resolution of dependencies too? I'm not sure if that would potentially yield more taps and therefore be necessary to make my chance catch all cases.

I think just consistency is the main thing. It's not clear as-is that `kept_formula` and `formulae_to_uninstall` will be consistent which is not ideal.

> How much do you want the new method to include? All the resolution of dependencies too? I'm not sure if that would potentially yield more taps and therefore be necessary to make my chance catch all cases.

Up to you. As above: I think consistency is the main thing to aim for.

> `@dsl.entries.select { |e| e.type == :brew }.map(&:name)` could arguably be a helper method on `Bundle::Dsl` also.

Leave it in this file for this PR, thanks.